src/websocket/server.cc

```diff
@@ -257,16 +257,28 @@ future<websocket_parser::consumption_result_t> websocket_parser::operator()(
         }
     }
     if (_state == parsing_state::payload) {
-        if (_payload_length > data.size()) {
-            _payload_length -= data.size();
-            remove_mask(data, data.size());
-            _result = std::move(data);
-            return websocket_parser::stop(buff_t(0));
+        if (data.size() < _remaining_payload_length) {
+            if (_result.empty()) {
+                _parsed_payload_length = 0;
+                _result = temporary_buffer<char>(_remaining_payload_length);
+            }
+            std::copy(data.begin(), data.end(), _result.get_write() + _parsed_payload_length);
+            _parsed_payload_length += data.size();
+            _remaining_payload_length -= data.size();
+            return websocket_parser::dont_stop();
         } else {
-            _result = data.clone();
-            remove_mask(_result, _payload_length);
-            data.trim_front(_payload_length);
-            _payload_length = 0;
+            if (_result.empty()) {
+                _result = data.clone();
```

Comment:
Next patch changes this to be `data.share()`. Can this be share()d right here?

Reply:
I didn't want to mix too many changes in a single commit. If there is a regression, then it can be bisected with more granularity.